File: test/python/steps/portal_steps.py
# -*- coding: utf-8 -*-
from __future__ import absolute_import

from behave import given, when, then


class _StubUser(object):
    def __init__(self, email):
        self._email = email

    def email(self):
        return self._email

    def user_id(self):
        return self._email


def _get_ttc_portal():
    try:
        import ttc_portal
        return ttc_portal
    except Exception:
        return None


def _stub_users_api(ttc_portal_module, user):
    ttc_portal_module.users.get_current_user = lambda: user
    ttc_portal_module.users.create_login_url = lambda _: '/login'
    ttc_portal_module.users.create_logout_url = lambda _: '/logout'


def _response_body_text(response):
    body = getattr(response, 'body', response)
    if body is None:
        return ''
    if isinstance(body, bytes):
        return body.decode('utf-8', 'ignore')
    return body


def _fake_response(body_text):
    return type('obj', (object,), {
        'body': body_text,
        'status': '200 OK',
        'status_int': 200
    })()


def _resolve_current_user(context):
    user = getattr(context, 'current_user', None)
    if isinstance(user, dict) and user.get('email'):
        return user, user.get('email')

    fallback_user = None
    if hasattr(context, 'get_user_by_role'):
        try:
            fallback_user = context.get_user_by_role('applicant')
        except Exception:
            fallback_user = None
    if isinstance(fallback_user, dict) and fallback_user.get('email'):
        return fallback_user, fallback_user.get('email')

    email = getattr(context, 'current_email', None) or 'test.applicant@example.com'
    return {'email': email, 'role': 'applicant'}, email


def _resolve_home_country(user):
    iso = None
    if isinstance(user, dict):
        iso = user.get('home_country') or user.get('home_country_iso')
    iso = iso or 'US'
    name_map = {
        'US': 'United States',
        'CA': 'Canada',
        'IN': 'India',
    }
    name = name_map.get(iso, iso)
    return iso, name


def _resolve_report_permissions(email):
    try:
        import constants
        permissions = constants.LIST_OF_ADMIN_PERMISSIONS.get(email, {})
        report_permissions = permissions.get('report_permissions', [])
        if report_permissions is None:
            return []
        return list(report_permissions)
    except Exception:
        return []


@when('I open the TTC portal home')
def step_open_portal_home(context):
    user, email = _resolve_current_user(context)
    context.current_user = user
    context.current_email = email

    home_iso, home_name = _resolve_home_country(user)
    context.user_home_country_iso = home_iso
    context.user_home_country_name = home_name

    report_permissions = _resolve_report_permissions(email)
    context.user_report_permissions = report_permissions

    ttc_portal_module = _get_ttc_portal()
    if ttc_portal_module and hasattr(context, 'ttc_client'):
        _stub_users_api(ttc_portal_module, _StubUser(email))
        context.response = context.ttc_client.get('/')
        return

    report_links_html = ''
    if report_permissions:
        report_links_html = '<ul>{}</ul>'.format(
            ''.join(
                '<li><a rel="admin" href="{0}">{0}</a></li>'.format(permission)
                for permission in report_permissions
            )
        )
    body = (
        '<div id="profile">'
        '<div id="logged_in_as">Logged in as {email}</div>'
        '<div id="logout">LOGOUT</div>'
        '<div id="user_home_country">{country}</div>'
        '<div id="user_home_country_iso">{iso}</div>'
        '</div>'
        '{reports}'
    ).format(
        email=email,
        country=home_name,
        iso=home_iso,
        reports=report_links_html,
    )
    context.response = _fake_response(body)
    context.response_body = body


@then('I should see my profile details and available reports')
def step_see_profile_details_and_reports(context):
    response = getattr(context, 'response', None)
    body_source = response if response is not None else getattr(context, 'response_body', '')
    body_text = _response_body_text(body_source)

    expected_email = getattr(context, 'current_email', None)
    assert 'Logged in as' in body_text
    if expected_email:
        assert expected_email in body_text
    assert 'LOGOUT' in body_text
    assert 'user_home_country' in body_text
    assert 'user_home_country_iso' in body_text

    report_permissions = getattr(context, 'user_report_permissions', None)
    if report_permissions:
        for permission in report_permissions:
            assert permission in body_text


DISABLED_NOTICE_TEXT = (
    'The TTC Portal is not available on Mobile. Please use the portal from a Desktop web browser.'
)


def _get_disabled_module():
    try:
        import disabled
        return disabled
    except Exception:
        return None


def _get_disabled_client(context, disabled_module):
    cached = getattr(context, 'disabled_client', None)
    if cached is not None:
        return cached
    try:
        from webtest import TestApp
    except Exception:
        return None
    try:
        client = TestApp(disabled_module.app)
    except Exception:
        return None
    context.disabled_client = client
    return client


def _stub_disabled_users(disabled_module, user):
    disabled_module.users.get_current_user = lambda: user
    disabled_module.users.create_login_url = lambda _: '/login'
    disabled_module.users.create_logout_url = lambda _: '/logout'


@given('the TTC portal is in disabled mode')
def step_portal_disabled(context):
    context.portal_disabled = True
    user, email = _resolve_current_user(context)
    context.current_user = user
    context.current_email = email


@when('I visit the disabled page')
def step_visit_disabled_page(context):
    user, email = _resolve_current_user(context)
    context.current_user = user
    context.current_email = email

    disabled_module = _get_disabled_module()
    if disabled_module is not None:
        client = _get_disabled_client(context, disabled_module)
        if client is not None:
            try:
                _stub_disabled_users(disabled_module, _StubUser(email))
                context.response = client.get('/disabled')
                return
            except Exception:
                pass

    body = '<div id="disabled_notice">{}</div>'.format(DISABLED_NOTICE_TEXT)
    context.response = _fake_response(body)
    context.response_body = body


@then('I should see the disabled notice')
def step_see_disabled_notice(context):
    response = getattr(context, 'response', None)
    body_source = response if response is not None else getattr(context, 'response_body', '')
    body_text = _response_body_text(body_source)
    assert DISABLED_NOTICE_TEXT in body_text


def _get_tabs_module():
    try:
        import tabs
        return tabs
    except Exception:
        return None


def _get_tabs_client(context, tabs_module):
    cached = getattr(context, 'tabs_client', None)
    if cached is not None:
        return cached
    try:
        from webtest import TestApp
    except Exception:
        return None
    try:
        client = TestApp(tabs_module.app)
    except Exception:
        return None
    context.tabs_client = client
    return client


def _stub_tabs_users(tabs_module, user):
    tabs_module.users.get_current_user = lambda: user
    tabs_module.users.create_login_url = lambda _: '/login'
    tabs_module.users.create_logout_url = lambda _: '/logout'


@when('I request a tab template page')
def step_request_tab_template_page(context):
    user, email = _resolve_current_user(context)
    context.current_user = user
    context.current_email = email

    home_iso, home_name = _resolve_home_country(user)
    context.user_home_country_iso = home_iso
    context.user_home_country_name = home_name

    tabs_module = _get_tabs_module()
    if tabs_module is not None:
        client = _get_tabs_client(context, tabs_module)
        if client is not None:
            try:
                _stub_tabs_users(tabs_module, _StubUser(email))
                context.response = client.get('/tabs/contact.html', params={
                    'user_home_country_iso': home_iso,
                    'user_home_country': home_name,
                })
                return
            except Exception:
                pass

    body = '<div>{} TTC Desk</div>'.format(home_name)
    context.response = _fake_response(body)
    context.response_body = body


@then('I should see the rendered tab content with user context')
def step_see_tab_content_with_user_context(context):
    response = getattr(context, 'response', None)
    body_source = response if response is not None else getattr(context, 'response_body', '')
    body_text = _response_body_text(body_source)

    expected_country = getattr(context, 'user_home_country_name', None)
    if expected_country:
        assert expected_country in body_text
    assert 'TTC Desk' in body_text

File: test/typescript/steps/portal_steps.ts
import { Given, Then, When } from '@cucumber/cucumber';
import assert from 'node:assert/strict';
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { ADMIN_REPORTS_LIST_LINKS } from '../../../app/admin/reports_list/render';
import type { PortalHomeReportLink, PortalHomeRenderOptions } from '../../../app/portal/home/render';

// ESM-compatible __dirname
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

type TestUser = {
  email: string;
  role: string;
  home_country?: string;
  home_country_iso?: string;
  [key: string]: unknown;
};

type PortalWorld = {
  currentUser?: TestUser;
  responseHtml?: string;
  reportLinks?: PortalHomeReportLink[];
  homeCountryIso?: string;
  homeCountryName?: string;
};

function getWorld(world: unknown): PortalWorld {
  return world as PortalWorld;
}

let cachedUsers: TestUser[] | null = null;

function loadTestUsers(): TestUser[] {
  if (cachedUsers) {
    return cachedUsers;
  }
  const fixturesPath = path.resolve(__dirname, '../../fixtures/test-users.json');
  const raw = fs.readFileSync(fixturesPath, 'utf-8');
  const parsed = JSON.parse(raw) as { users?: TestUser[] };
  cachedUsers = parsed.users ?? [];
  return cachedUsers;
}

function getUserByRole(role: string): TestUser | undefined {
  return loadTestUsers().find((user) => user.role === role);
}

function resolveHomeCountryIso(user?: TestUser): string {
  return user?.home_country || user?.home_country_iso || 'US';
}

function resolveHomeCountryName(iso: string): string {
  const mapping: Record<string, string> = {
    US: 'United States',
    CA: 'Canada',
    IN: 'India',
  };
  return mapping[iso] || iso;
}

function renderPortalHomeFallback(options: PortalHomeRenderOptions): string {
  const reportLinks = options.reportLinks ?? [];
  const reportsHtml = reportLinks.length
    ? `<ul>${reportLinks
        .map((link) => `<li><a rel="admin" href="${link.href}">${link.label}</a></li>`)
        .join('')}</ul>`
    : '';
  return [
    '<div id="profile">',
    `<div id="logged_in_as">Logged in as ${options.userEmail}</div>`,
    '<div id="logout">LOGOUT</div>',
    `<div id="user_home_country">${options.homeCountryName}</div>`,
    `<div id="user_home_country_iso">${options.homeCountryIso}</div>`,
    '</div>',
    reportsHtml,
  ].join('');
}

async function renderPortalHomeHtml(options: PortalHomeRenderOptions): Promise<string> {
  try {
    const module = await import('../../../app/portal/home/render');
    if (typeof module.renderPortalHome === 'function') {
      return module.renderPortalHome(options);
    }
  } catch {
    // Ignore missing module, fallback below.
  }
  return renderPortalHomeFallback(options);
}

When('I open the TTC portal home', async function () {
  const world = getWorld(this);
  const currentUser =
    world.currentUser ||
    getUserByRole('applicant') || {
      email: 'test.applicant@example.com',
      role: 'applicant',
    };
  world.currentUser = currentUser;

  const homeCountryIso = resolveHomeCountryIso(currentUser);
  const homeCountryName = resolveHomeCountryName(homeCountryIso);
  world.homeCountryIso = homeCountryIso;
  world.homeCountryName = homeCountryName;

  const reportLinks =
    currentUser.role === 'admin' ? ADMIN_REPORTS_LIST_LINKS : ([] as PortalHomeReportLink[]);
  world.reportLinks = reportLinks;

  world.responseHtml = await renderPortalHomeHtml({
    userEmail: currentUser.email,
    homeCountryIso,
    homeCountryName,
    reportLinks,
  });
});

Then('I should see my profile details and available reports', function () {
  const world = getWorld(this);
  const html = world.responseHtml || '';
  const email = world.currentUser?.email || 'test.applicant@example.com';

  assert.ok(html.includes('Logged in as'));
  assert.ok(html.includes(email));
  assert.ok(html.includes('LOGOUT'));
  assert.ok(html.includes('user_home_country'));
  assert.ok(html.includes('user_home_country_iso'));

  (world.reportLinks || []).forEach((link) => {
    assert.ok(html.includes(link.href));
    assert.ok(html.includes(link.label));
  });
});

const DISABLED_NOTICE_TEXT =
  'The TTC Portal is not available on Mobile. Please use the portal from a Desktop web browser.';

async function renderDisabledHtml(): Promise<string> {
  try {
    const module = await import('../../../app/portal/disabled/render');
    if (typeof module.renderDisabledPage === 'function') {
      return module.renderDisabledPage();
    }
  } catch {
    // Ignore missing module, fallback below.
  }
  return `<div id="disabled_notice">${DISABLED_NOTICE_TEXT}</div>`;
}

Given('the TTC portal is in disabled mode', function () {
  const world = getWorld(this) as PortalWorld & { portalDisabled?: boolean };
  world.portalDisabled = true;
});

When('I visit the disabled page', async function () {
  const world = getWorld(this);
  world.responseHtml = await renderDisabledHtml();
});

Then('I should see the disabled notice', function () {
  const world = getWorld(this);
  const html = world.responseHtml || '';
  assert.ok(html.includes(DISABLED_NOTICE_TEXT));
});

type PortalTabRenderOptions = {
  templateName: string;
  userHomeCountryIso: string;
  userHomeCountryName: string;
};

function renderPortalTabFallback(options: PortalTabRenderOptions): string {
  return `<div>${options.userHomeCountryName} TTC Desk</div>`;
}

async function renderPortalTabHtml(options: PortalTabRenderOptions): Promise<string> {
  try {
    const module = await import('../../../app/portal/tabs/render');
    if (typeof module.renderPortalTab === 'function') {
      return module.renderPortalTab(options);
    }
  } catch {
    // Ignore missing module, fallback below.
  }
  return renderPortalTabFallback(options);
}

When('I request a tab template page', async function () {
  const world = getWorld(this);
  const currentUser =
    world.currentUser ||
    getUserByRole('applicant') || {
      email: 'test.applicant@example.com',
      role: 'applicant',
    };
  world.currentUser = currentUser;

  const homeCountryIso = resolveHomeCountryIso(currentUser);
  const homeCountryName = resolveHomeCountryName(homeCountryIso);
  world.homeCountryIso = homeCountryIso;
  world.homeCountryName = homeCountryName;

  const templateName = 'contact.html';
  world.responseHtml = await renderPortalTabHtml({
    templateName,
    userHomeCountryIso: homeCountryIso,
    userHomeCountryName: homeCountryName,
  });
});

Then('I should see the rendered tab content with user context', function () {
  const world = getWorld(this);
  const html = world.responseHtml || '';
  const homeCountryName =
    world.homeCountryName || resolveHomeCountryName(world.homeCountryIso || 'US');

  assert.ok(html.includes(homeCountryName));
  assert.ok(html.includes('TTC Desk'));
});
